ir/expression_ref.h

```diff
@@ -0,0 +1,82 @@
+#ifndef IR_EXPRESSION_REF_H_
+#define IR_EXPRESSION_REF_H_
+
+#include "ir.h"
+
+namespace IR {
+
+class ERef {
+    const Expression *expr;
+
+    template<class T> class ERef_specific {
+        const T *expr;
+        friend class ERef;
+
+     public:
+        ERef_specific(const T *e) : expr(e) {}  // NOLINT(runtime/explicit)
+        const T &operator*() const { return *expr; }
+        const T *operator->() const { return expr; }
+        operator const T *() const { return expr; }
+    };
+
+ public:
+    ERef(const Expression *e) : expr(e) {}   // NOLINT(runtime/explicit)
+    ERef(const Expression &e) : expr(&e) {}   // NOLINT(runtime/explicit)
```

Comment:
Yes, they should always be allocated by new.  This was an attempt to get around the need for explicit `IR::ERef` constructor calls all over the place, but it doesn't really work.